[docs] Add custom builds config schema (#22165)
expo/expo: docs/pages/preview/custom-build-config-schema.mdx

## YAML syntax for workflows

Workflow files are stored inside the **.eas/build** directory path. They use YAML syntax and must have a `.yml` or `.yaml` file extension. If you are new to YAML or want to learn more about the syntax, see [Learn YAML in Y minutes](https://learnxinyminutes.com/docs/yaml/).

## `build`

Defined to describe a custom workflow. All config options to create a workflow are specified under it.

### `name`

The name of your workflow that is used to identify the workflow in the build logs. EAS Build uses this property to display the name of your workflow in the dashboard.

For example, the workflow's name is "Run tests":

```yaml
build:
  # @info #
  name: Run tests
  # @end #
  steps:
    - run:
        name: Install dependencies
        command: npm install
```

### `steps`

Steps are used to describe a list of actions either in the form of commands or functions. These actions are executed when a workflow is triggered. You can define single or multiple steps in a workflow. It is **required** to define at least one step per workflow.

Each step is configured with the following properties.

#### `steps.run`

> The `run` key is used to trigger a set of instructions. For example, a `run` key is used to install dependencies using the `npm install` command:

Author: Yeah, I don't think the example below or the definition above for `run` signifies that. I think, since you are pointing this out, it might be important for the specific use case to showcase that there is a possibility for the end-developer to do so.

Do you think updating the example below as you suggested will be a good option here? Because, in `steps.command` we are already specifying that the executing multiline bash scripts is possible. Initially, I didn't want to include this specific phrase in the `run` because I think running multiline bash scripts isn't possible without using `command` key (correct me if my assumption is wrong here).

```yaml
build:
  name: Install npm dependencies
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    # @end #
```

You can also `steps.run` to execute inline bash shell commands:

```yaml
build:
  name: Run inline shell commands
  steps:
    # @info #
    - run: echo "Hello world"
    # @end #
```

#### `steps.name`

The name of a step that is used in a `build` to identify which step triggers in the build logs.

#### `steps.command`

The `command` is defins a custom shell command to run when a step is executed. It is **required** to define a command for each step and can be a multiline bash shell command.

#### `steps.working_directory`

The `working_directory` is used to define an existing directory from the project's root directory. For example, a step is created to list all the assets inside the **assets** directory, which is a directory in your Expo project. The `working_directory` is set to `assets`:

```yaml
build:
  name: Demo
  steps:
    - run:
        name: List assets
        # @info #
        working_directory: assets
        # @end #
        command: ls -la
```

> 

Reviewer: Also, it's possible to define `inputs` and `outputs`.

#### Use a single step

For example, a workflow with the following `steps` will run a single command to install npm dependencies:

```yaml
build:
  name: Install npm dependencies
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    # @end #
```

> **info** `-` before `run` counts as indentation.

#### Use multiple steps

When multiple `steps` are defined, they are executed sequentially. For example, a workflow with the following `steps` will first run a single command to install npm dependencies and then run a command to run tests:

```yaml
build:
  name: Run tests
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
    # @end #
```

## `functions`

Defined to describe a reusable function or [built-in EAS functions](#built-in-eas-functions) that can be used in a workflow. All config options to create a function are specified with the following properties:

> ### `name`

Reviewer: Please apply the same suggestion to the other function properties.
```suggestion
### `functions.FUNCTION_NAME.name`
```

Reviewer: Also, before describing `.name`, I'd add a small note about the `FUNCTION_NAME`. I mean that this is an identifier that you use in `build.steps` to call the function

The name of a function that is used in a `build`. It is also used to identify the function triggered in the build logs. For example, a function named `greetings` is created:

```yaml
functions:
  greetings:
    # @info #
    name: Say Hi!
    # @end #
```

### `platforms`

> Name of the platform that a specific function should get executed. `linux` is used for Android and `darwin` for iOS. If not specified, the function is executed for both platforms when the appropriate build runs for each platform.

Reviewer: It's more of a platform check. The list specifies the platforms that the function is compatible with. Anyway, let's comment this out for now. It's not implemented.

### `inputs`

> An input value is provided to a function.

Reviewer: ```suggestion
Input values are provided to a function.
```

Reviewer: A function can have multiple inputs, see https://github.com/expo/eas-custom-builds-example/blob/main/.eas/build/functions.yml#L23

> 

Author: @dsokal, does `.name` and `.required` also apply to `steps[].input/output`?

> 

Reviewer: - `steps[].inputs` is only a key-value map:
  ```
    - run:
        name: Use output from another step
        inputs:
          foo: Literal
          bar: ${ steps.id456.required_param }
          baz: ${ steps.id456.optional_param }
        command: echo "${ inputs.foo} ${ inputs.bar} ${ inputs.baz}"
  ```
- `steps[].outputs` can have `.name` and `.required`:
  ```
    - run:
        name: Produce another output
        id: id456
        outputs:
          - required_param
          - name: optional_param
            required: false
        command: |
          echo "Producing more output"
          set-output required_param "abc 123 456"
  ```

#### `inputs.default_value`

You can use `default_value` to provide one default input. For example, a function has a default value of `Hello world`:

```yaml
functions:
  greetings:
    name: Say Hi!
    inputs:
      - name: name
        # @info #
        default_value: Hello world
        # @end #
```

#### `inputs.allowed_values`

You can use `allowed_values` to provide multiple values in an array. For example, a function has multiple allowed values:

```yaml
functions:
  greetings:
    name: Say Hi!
    inputs:
      - name: name
        default_value: Hello world
        # @info #
        allowed_values: [Hi, Hello, Hey]
        # @end #
```

### `outputs`

An output value is expected from a function. For example, a function has an output value of `Hello world`:

```yaml
functions:
  greetings:
    name: Say Hi!
    # @info #
    outputs: [value]
    # @end #

>     command: set-output value Hello

Reviewer: If I understand the line above (https://github.com/expo/expo/pull/22165/files#diff-7e91ba88739076f4e473a23a6f5413419ac30c5694ba4fe3d0fc9e84c52c496fR184) correctly, this has to be:
```suggestion
    command: set-output value "Hello world"
```

```

> **info** Any function that is built-in and provided by EAS must start with the prefix `eas/`.

The `eas/upload_artifact` function has the following properties

| Property      | Description                                                                 |
| ------------- | --------------------------------------------------------------------------- |
| `name`        | The name of the step in the reusable function that shows in the build logs. |
| `inputs`      | Requires input providers (such as `path` or `type`) to be defined.          |
| `inputs.path` | **Required**. The path to the artifact that is uploaded to the EAS servers. |
| `inputs.type` | The type of artifact that is uploaded to the EAS servers.                   |

## Use a reusable function in a `build`

For example, a workflow with the following reusable function contains a single command to print a message that is echoed.

```yaml
functions:
  greetings:
    - name: name
      default_value: Hello world
    inputs: [value]
    command: echo "${ inputs.name }, { inputs.value }"
```

The above function can be used in a `build` as follows:

```yaml
build:
  name: Functions Demo
  steps:
    # @info #

>     - greetings:

Reviewer: Maybe it's worth describing that you can override `working_directory`, `name`, and `shell`.

Author: Can yuo give me an example how overriding any of those value is possible? Or should we just add that as a note?

Reviewer: Let's say that you have a reusable function for listing files in a directory:
```
functions:
  list_files:
    name: List files
    command: ls -la
```
Now, if you use it in your workflow:
```
build:
  name: List files
  steps:
    - list_files
```
It'll list files in the root directory of your project.

You can change this behavior by overriding `working_directory` in the function call:
```
build:
  name: List files
  steps:
    - list_files:
         working_directory: /a/b/c
```

```yaml
build:
  name: Install npm dependencies
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    # @end #
```

You can also use `steps.run` to execute inline bash shell commands:

```yaml
build:
  name: Run inline shell commands
  steps:
    # @info #
    - run: echo "Hello world"
    # @end #
```

#### Use a single step

For example, a workflow with the following `steps` will run a single command to install npm dependencies:

```yaml
build:
  name: Install npm dependencies
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    # @end #
```

> **Note:** `-` before `run` counts as indentation.

#### Use multiple steps

When multiple `steps` are defined, they are executed sequentially. For example, a workflow with the following `steps` will first run a single command to install npm dependencies and then run a command to run tests:

```yaml
build:
  name: Run tests
  steps:
    # @info #
    - run:
        name: Install dependencies
        command: npm install
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
    # @end #
```

> 

Reviewer: Please add a section about `steps[].run.id`.

There are two use cases for that: 
1. Calling the same function that produces an output/outputs multiple times.
2. Using output from a step (not a function call!) in another step.

(1)
Let's say you have a function for generating a random value:
```
  random:
    name: Generate random number
    outputs: [value]
    command: set-output value `random_number`
```
You want to generate two random number and print them in a step:
```

build:
  name: Functions Demo
  steps:
    - random:
        id: random_1
    - random:
        id: random_2
    - run:
        name: Print random numbers
        inputs:
          random_1: ${ steps.random_1.value }
          random_2: ${ steps.random_2.value }
        command: | 
          echo "${ inputs.random_1 }"
          echo "${ inputs.random_2 }"
```

(2)
You have a step that produces an output and you want to access it in another step:

```
    - run:
        name: Produce output
        id: id123    # <---- !!!
        outputs: [foo]
        command: |
          echo "Producing output for another step"
          set-output foo bar
    - run:
        name: Use output from another step
        inputs:
          foo: ${ steps.id123.foo }
        command: |
          echo "foo = \"${ inputs.foo }\""
```

```

#### `inputs[].allowed_values`

You can use `allowed_values` to provide multiple values in an array. For example, a function has multiple allowed values:

```yaml
functions:
  greetings:
    name: Say Hi!
    inputs:
      - name: name
        default_value: Hello world
        # @info #
        allowed_values: [Hi, Hello, Hey]
        # @end #
```

#### Multiple input values

Multiple input values can be provided to a function.

```yaml
functions:
  greetings:
    name: Say Hi!
    # @info #
    inputs:
      - name: name
        default_value: Expo
      - name: greeting
        default_value: Hi
        allowed_values: [Hi, Hello]
    # @end #
    command: echo "${ inputs.greeting }, ${ inputs.name }!"
```

### `function.function_name.outputs`

An output value is expected from a function. For example, a function has an output value of `Hello world`:

```yaml
functions:
  greetings:
    name: Say Hi!
    # @info #
    outputs: [value]
    # @end #
    command: set-output value "Hello world"
```

#### `outputs[].name`

The name of the output value. It is used as an identifier to access the output value in another step:

```yaml
functions:
  greetings:
    name: Say Hi!

>     inputs:

Reviewer: ```suggestion
    outputs:
```

      # @info #
      - name: name
      # @end #
```

#### `outputs[].required`

Boolean to indicate if the output value is required or not. For example, a function does not have a required output value:

```yaml
functions:
  greetings:
    name: Say Hi!
    outputs:
      - name: value
        # @info #
        required: false
        # @end #
```

### `function.function_name.command`

Used to define the command to run when a function is executed. It is **required** to define a command for each reusable function. For example, the command `echo "Hello world"` is used to print a message:

```yaml
functions:
  greetings:
    name: Say Hi!
    # @info #
    command: echo "Hello world"
    # @end #
```

### `function.function.name.shell`

> Used to define the default executable shell for a system. For example, the default shell is set to the system's shell:

Reviewer: ```suggestion
Used to define the default executable shell for a step. For example, the step's shell is set to `/bin/sh`:
```

          echo "foo = \"${ inputs.foo }\""
```

## `functions`

Defined to describe a reusable function that can be used in a workflow. All config options to create a function are specified with the following properties:

### `functions.[function_name]`

The `[function_name]` is the name of a function that you define to identify it in the `build.steps`. For example, you can define a function with the name `greetings`:

{/* prettier-ignore */}
```yaml
functions:
  # @info #
  greetings:
  # @end #
    name: Say Hi!
```

### `functions.[function_name].name`

The name that is used in build logs to display the name of the function. For example, a function with the display name `Say Hi!`:

```yaml
functions:
  greetings:
    # @info #
    name: Say Hi!
    # @end #
```

{/* Note: (aman) Commenting this out since it's not implemented yet. */}
{/* `platforms` */}
{/* Name of the platform that a specific function should get executed. `linux` is used for Android and `darwin` for iOS. If not specified, the function is executed for both platforms when the appropriate build runs for each platform. */}

### `functions.[function_name].inputs`

Input values are provided to a function.

> #### `functions.[function_name].inputs[].name`

Reviewer: To match the sections below:
```suggestion
#### `inputs[].name`
```
